### PR TITLE
Partition libs

### PR DESCRIPTION
**Pull request by larryk85**
The PR partitions eosiolib into Core, Contracts and CAPI.  To ensure backwards compatibility, the whole of eosiolib is still accessible from `#include <eosiolib/header_file.h/hpp>`.  If you include one of these header files you will be given a warning stating that that particular header file is deprecated and what to replace it with.  For the header files in CAPI, an additional message is given stating that these will no longer be accessible from a C++ smart contract after v1.7.0.  All uses of the CAPI in the C++ API are namespaced with `internal_use_do_not_use` this should be a very clear indicator to users that the CAPI should not be accessed.  Furthermore,  only certain groups of header files are accessible given particular compilation modes.   The CAPI is only accessible now through `eosio-cc`, the Contracts API will only be available to `eosio-cpp` compiling for EOSIO `smart contracts` and Core API is available to any "target",  `eosio-cpp` in native mode, i.e. unit tests, will have access to all current and future API groups.  Going forward instead of `#include <eosiolib/header.hpp>`, the pattern will be `#include <eosio/header.hpp>`, this is the same for the native tester, which will now be `#include <eosio/tester.hpp>` instead of `#include <eosio/native/tester.hpp>`.

Division of APIs
### CAPI ###
- `action.h`
- `chain.h`
- `crypto.h`
- `db.h`
- `permission.h`
- `print.h`
- `privileged.h`
- `system.h`
- `transaction.h`
- `types.h`

### Core ###
- `asset.hpp`
- `binary_extension.hpp`
- `crypto.hpp`
- `datastream.hpp`
- `fixed_bytes.hpp`
- `ignore.hpp`
- `name.hpp`
- `print.hpp`
- `rope.hpp`
- `serialize.hpp`
- `symbol.hpp`
- `system.hpp`
- `time.hpp`
- `varint.hpp`

### Contracts ###
- `action.hpp`
- `contract.hpp`
- `dispatcher.hpp`
- `eosio.hpp`
- `multi_index.hpp`
- `permission.hpp`
- `privileged.hpp`
- `producer_schedule.hpp`
- `singleton.hpp`
- `transaction.hpp`

### CODE REVIEW

File: libraries/eosiolib/capi/eosio/action.h [hunk 1]
```diff
@@ -0,0 +1,174 @@
+/**
+ *  @file
+ *  @copyright defined in eos/LICENSE
+ */
+#pragma once
+#include "system.h"
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/capi/eosio/action.h]: got it.

File: libraries/eosiolib/contracts/eosio/permission.hpp [hunk 2]
```diff
@@ -30,6 +30,49 @@ namespace eosio {
...
+   *  @param delay_us - the provided delay in microseconds (cannot exceed INT64_MAX)
+   *
+   *  @return 1 if the permission is authorized, 0 otherwise
+   */
+   bool
+   check_permission_authorization( name account,
+                                   name permission,
+                                   const char* pubkeys_data, uint32_t pubkeys_size,
+                                   const char* perms_data,   uint32_t perms_size,
+                                   uint64_t delay_us ) {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/permission.hpp]: got it.

File: libraries/eosiolib/core/eosio/ignore.hpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    *  Deserialize an ignored type from a stream
+    *
+    *  @brief Deserialize an ignored type
+    *  @param ds - The stream to read
+    *  @param ignored - The destination for deserialized value
+    *  @tparam DataStream - Type of datastream buffer
+    *  @return DataStream& - Reference to the datastream
+    */
+   template<typename DataStream, typename T>
+   inline DataStream& operator>>(DataStream& ds, ::eosio::ignore<T>) {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/core/eosio/ignore.hpp]: got it.

File: libraries/eosiolib/core/eosio/print.hpp [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+   inline void printl( const char* ptr, size_t len ) {
+     internal_use_do_not_use::prints_l(ptr, len);
+   }
+
+   /**
+    *  Prints string
+    *
+    *  @param ptr - a null terminated string
+    */
+   inline void print( const char* ptr ) {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/core/eosio/print.hpp]: got it.

File: libraries/eosiolib/core/eosio/time.hpp [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+   inline time_point current_time_point() {
+     return time_point(current_time());
+   }
+
+   /**
+    *  Returns the time in microseconds from 1970 of the current block as a block_timestamp
+    *
+    *  @return time in microseconds from 1970 of the current block as a block_timestamp
+    */
+   inline block_timestamp current_block_time() {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/core/eosio/time.hpp]: Will do.

File: libraries/eosiolib/core/eosio/time.hpp [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+
+   typedef block_timestamp block_timestamp_type;
+
+   /**
+    *  Returns the time in microseconds from 1970 of the current block
+    *
+    *  @return time in microseconds from 1970 of the current block
+    */
+   inline microseconds current_time() {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/core/eosio/time.hpp]: got it.

File: libraries/eosiolib/core/eosio/time.hpp [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+   inline block_timestamp current_block_time() {
+     return block_timestamp(current_time_point());
+   }
+
+    /**
+    *  Get time (rounded down to the nearest second) of the current block (i.e. the block including this action)
+    *
+    *  @return time in seconds from 1970 of the current block
+    */
+  inline uint32_t  now() {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/core/eosio/time.hpp]: got it.

File: libraries/eosiolib/core/eosio/system.hpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    *  Example:
+    *
+    *  @code
+    *  eosio_exit(0);
+    *  eosio_exit(1);
+    *  eosio_exit(2);
+    *  eosio_exit(3);
+    *  @endcode
+    */
+   inline void eosio_exit( int32_t code ) {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/core/eosio/system.hpp]: got it.

File: libraries/eosiolib/core/eosio/varint.hpp [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+       /**
+        *  Deserialize an signed_int object
+        *
+        *  @param ds - The stream to read
+        *  @param vi - The destination for deserialized value
+        *  @tparam DataStream - Type of datastream
+        *  @return DataStream& - Reference to the datastream
+        */
+       template<typename DataStream>
+       friend DataStream& operator >> ( DataStream& ds, signed_int& vi ){
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/core/eosio/varint.hpp]: got it.

File: libraries/eosiolib/contracts/eosio/permission.hpp [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    *  @param account    - the account owner of the permission
+    *  @param permission - the permission name to check for authorization
+    *  @param provided_keys - the set of public keys which have authorized the transaction
+    *  @param provided_permissions - the set of permissions which have authorized the transaction (empty permission name acts as wildcard)
+    *  @param provided_delay_us - the provided delay in microseconds (cannot exceed INT64_MAX)
+    *
+    *  @return whether the permission was authorized by provided delay, keys, and permissions
+    */
+   bool
+   check_permission_authorization( name                               account,
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/permission.hpp]: will do in the next release.

File: libraries/eosiolib/contracts/eosio/privileged.hpp [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+   }
+
+   /**
+    * Set new active producers. Producers will only be activated once the block which starts the next round is irrreversible
+    *
+    * @param producer_data - pointer to producer schedule packed as bytes
+    * @param producer_data_size - size of the packed producer schedule
+    * @pre `producer_data` is a valid pointer to a range of memory at least `producer_data_size` bytes long that contains serialized produced schedule data
+    */
+   inline void set_active_producers( char *producer_data, uint32_t producer_data_size ) {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/privileged.hpp]: got it.

File: libraries/eosiolib/contracts/eosio/privileged.hpp [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+   /**
+    * Proposes a schedule change
+    *
+    * @note Once the block that contains the proposal becomes irreversible, the schedule is promoted to "pending" automatically. Once the block that promotes the schedule is irreversible, the schedule will become "active"
+    * @param producer_data - packed data of produce_keys in the appropriate producer schedule order
+    * @param producer_data_size - size of the data buffer
+    *
+    * @return -1 if proposing a new producer schedule was unsuccessful, otherwise returns the version of the new proposed schedule
+    */
+   inline int64_t set_proposed_producers( char *producer_data, uint32_t producer_data_size ) {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/privileged.hpp]: got it.

File: libraries/eosiolib/contracts/eosio/producer_schedule.hpp [hunk 13]
```diff
@@ -0,0 +1,28 @@
+#pragma once
+#include "privileged.hpp"
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/producer_schedule.hpp]: got it.

File: libraries/eosiolib/contracts/eosio/producer_schedule.hpp [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+       */
+      uint32_t                     version;
+
+      /**
+       * List of producers for this schedule, including its signing key
+       */
+      std::vector<producer_key>    producers;
+   };
+
+   /// @} producer_schedule
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/producer_schedule.hpp]: got it.

File: libraries/eosiolib/contracts/eosio/transaction.hpp [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+     internal_use_do_not_use::send_deferred(sender_id, payer.value, serialized_transaction, size, replace);
+   }
+   /**
+    * Retrieve the indicated action from the active transaction.
+    *
+    * @param type - 0 for context free action, 1 for action
+    * @param index - the index of the requested action
+    * @return the indicated action
+    */
+   inline action get_action( uint32_t type, uint32_t index ) {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/transaction.hpp]: will do in the next release.

File: libraries/eosiolib/core/eosio/fixed_bytes.hpp [hunk 16]
```diff
@@ -89,80 +93,89 @@ namespace eosio {
...
-         * Constructor to fixed_key object from std::array of num_words() words
+         * @brief Constructor to fixed_bytes object from std::array of Word types smaller in size than word_t
          *
+         * @details Constructor to fixed_bytes object from std::array of Word types smaller in size than word_t
          * @param arr - Source data
          */
          template<typename Word, size_t NumWords,
                   typename Enable = typename std::enable_if<std::is_integral<Word>::value &&
                                                              !std::is_same<Word, bool>::value &&
                                                              sizeof(Word) < sizeof(word_t)>::type >
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/core/eosio/fixed_bytes.hpp]: oops.

File: libraries/eosiolib/core/eosio/fixed_bytes.hpp [hunk 17]
```diff
@@ -89,80 +93,89 @@ namespace eosio {
...
+         * @brief Constructor to fixed_bytes object from fixed-sized C array of Word types smaller in size than word_t
          *
-         * @details Constructor to fixed_key object from fixed-sized C array of Word types smaller in size than word_t
+         * @details Constructor to fixed_bytes object from fixed-sized C array of Word types smaller in size than word_t
          * @param arr - Source data
          */
          template<typename Word, size_t NumWords,
                   typename Enable = typename std::enable_if<std::is_integral<Word>::value &&
                                                              !std::is_same<Word, bool>::value &&
                                                              sizeof(Word) < sizeof(word_t)>::type >
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/core/eosio/fixed_bytes.hpp]: gotcha.

File: libraries/eosiolib/contracts/eosio/action.hpp [hunk 18]
```diff
@@ -0,0 +1,579 @@
...
+   inline time_point  publication_time() {
+     return time_point( microseconds ( internal_use_do_not_use::publication_time() ) );
+   }
+
+   /**
+   *  Get the current receiver of the action
+   *  @brief Get the current receiver of the action
+   *  @return the account which specifies the current receiver of the action
+   */
+   inline name current_receiver() {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/action.hpp]: Will do in the next release.

File: libraries/eosiolib/contracts/eosio/action.hpp [hunk 19]
```diff
@@ -0,0 +1,579 @@
...
+   inline void require_auth( name n ) {
+      internal_use_do_not_use::require_auth( n.value );
+   }
+
+   /**
+   *  Returns the time in microseconds from 1970 of the publication_time
+   *  @brief Get the publication time
+   *  @return the time in microseconds from 1970 of the publication_time
+   */
+   inline time_point  publication_time() {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/action.hpp]: Will do in the next release.

File: libraries/eosiolib/core/eosio/time.hpp [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+   inline microseconds current_time() {
+     return microseconds(static_cast<int64_t>(internal_use_do_not_use::current_time()));
+   }
+
+   /**
+    *  Returns the time in microseconds from 1970 of the current block as a time_point
+    *
+    *  @return time in microseconds from 1970 of the current block as a time_point
+    */
+   inline time_point current_time_point() {
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/core/eosio/time.hpp]: got it.

File: libraries/eosiolib/contracts/eosio/permission.hpp [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+    *  Checks if a transaction is authorized by a provided set of keys and permissions
+    *
+    *  @param trx - the transaction for which to check authorizations
+    *  @param provided_permissions - the set of permissions which have authorized the transaction (empty permission name acts as wildcard)
+    *  @param provided_keys - the set of public keys which have authorized the transaction
+    *
+    *  @return whether the transaction was authorized by provided keys and permissions
+    */
+   bool
+   check_transaction_authorization( const transaction&                 trx,
```
[the remark this answers is not in the document]
Comment (arhag) [reply libraries/eosiolib/contracts/eosio/permission.hpp]: I assume this is also something for the next release?

File: libraries/eosiolib/contracts/eosio/permission.hpp [hunk 22]
```diff
@@ -68,7 +69,7 @@ namespace eosio {
                                    name permission,
                                    const char* pubkeys_data, uint32_t pubkeys_size,
                                    const char* perms_data,   uint32_t perms_size,
-                                   uint64_t delay_us ) {
+                                   microseconds delay_us ) {
      internal_use_do_not_use::check_permission_authorization( account.value, permission.value, pubkeys_data, pubkeys_size, perms_data, perms_size, delay_us );
```
Comment (arhag) [inline libraries/eosiolib/contracts/eosio/permission.hpp]: I think the `delay_us` argument should be renamed to `delay` and the implementation should be changed to:
```cpp
int64_t delay_us = delay.count();
eosio::check( delay_us >= 0, "negative delay is not allowed" );
internal_use_do_not_use::check_permission_authorization( account.value, permission.value, pubkeys_data, pubkeys_size, perms_data, perms_size, static_cast<uint64_t>(delay_us) );
```

File: libraries/eosiolib/contracts/eosio/permission.hpp [hunk 23]
```diff
@@ -136,7 +137,7 @@ namespace eosio {
                                    name                               permission,
                                    const std::set<public_key>&        provided_keys,
                                    const std::set<permission_level>&  provided_permissions = std::set<permission_level>(),
-                                   uint64_t                           provided_delay_us = static_cast<uint64_t>(std::numeric_limits<int64_t>::max())
+                                   microseconds                       provided_delay_us = microseconds{std::numeric_limits<int64_t>::max()}
                                  )
    {
```
Comment (arhag) [inline libraries/eosiolib/contracts/eosio/permission.hpp]: I think the `provided_delay_us` argument should be renamed to `provided_delay` and the following two lines should be added to the start of this function:
```cpp
int64_t provided_delay_us = provided_delay.count();
eosio::check( provided_delay_us >= 0, "negative delay is not allowed" );
```

The usage of `provided_delay_us` below would also then need to be replaced with `static_cast<uint64_t>(provided_delay_us)`.

File: libraries/eosiolib/core/eosio/print.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (arhag) [reply libraries/eosiolib/core/eosio/print.hpp]: Turns out it is needed.

File: libraries/eosiolib/eosiolib.cpp [hunk 24]
```diff
@@ -1,20 +1,60 @@
...
+      return ct;
+   }
+
+   block_timestamp current_block_time() {
+      static auto bt = block_timestamp(current_time_point());
+      return bt;
+   }
+
+   std::vector<name> get_active_producers() {
+     static constexpr uint8_t prod_cnt = 21;
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/eosiolib.cpp]: got it.

File: libraries/eosiolib/eosiolib.cpp [hunk 25]
```diff
@@ -1,20 +1,60 @@
...
       char buf[sizeof(eosio::blockchain_parameters)];
-      size_t size = internal_use_do_not_use::get_blockchain_parameters_packed( buf, sizeof(buf) );
+      size_t size = get_blockchain_parameters_packed( buf, sizeof(buf) );
       eosio::check( size <= sizeof(buf), "buffer is too small" );
       eosio::datastream<const char*> ds( buf, size_t(size) );
       ds >> params;
    }
+
+   std::optional<uint64_t> set_proposed_producers( const std::vector<producer_key>& prods ) {
+      int64_t ret = set_proposed_producers((char*)prods.data(), prods.size()*sizeof(producer_key));
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/eosiolib.cpp]: got it.

File: libraries/eosiolib/contracts/eosio/privileged.hpp [hunk 26]
```diff
@@ -8,18 +9,13 @@ namespace eosio {
    namespace internal_use_do_not_use {
       extern "C" {
          __attribute__((eosio_wasm_import))
-         void get_resource_limits( uint64_t account, int64_t* ram_bytes, int64_t* net_weight, int64_t* cpu_weight );
-         __attribute__((eosio_wasm_import))
-         void set_resource_limits( uint64_t account, int64_t ram_bytes, int64_t net_weight, int64_t cpu_weight );
+         bool is_privileged( uint64_t account );
 
-         __attribute__((eosio_wasm_import))
-         int64_t set_proposed_producers( char *producer_data, uint32_t producer_data_size );
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/privileged.hpp]: got it.

File: libraries/eosiolib/contracts/eosio/permission.hpp (hunk 22 above)
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/permission.hpp]: got it.

File: libraries/eosiolib/contracts/eosio/permission.hpp (hunk 23 above)
Comment (larryk85) [reply libraries/eosiolib/contracts/eosio/permission.hpp]: got it.

**Pull request merged**